packages/storage-plus/src/deque.rs

```diff
@@ -0,0 +1,636 @@
+        storage
+            .get(&full_key)
+            .map(|vec| {
+                Ok(u32::from_be_bytes(
+                    vec.as_slice()
+                        .try_into()
+                        .map_err(|e| StdError::parse_err("u32", e))?,
+                ))
+            })
+            .unwrap_or(Ok(0))
+    }
+
+    /// Helper method for `set_tail` and `set_head` methods to write to storage
+    #[inline]
+    fn set_meta_key(&self, storage: &mut dyn Storage, key: &[u8], value: u32) {
+        let full_key = namespaces_with_key(&[self.namespace], key);
+        storage.set(&full_key, &value.to_be_bytes());
+    }
+
+    /// Returns the value at the given position in the queue or `None` if the index is out of bounds
+    pub fn get(&self, storage: &dyn Storage, pos: u32) -> StdResult<Option<T>> {
+        let head = self.head(storage)?;
+        let tail = self.tail(storage)?;
+
+        if pos >= calc_len(head, tail) {
+            // out of bounds
+            return Ok(None);
+        }
+
+        let pos = head.wrapping_add(pos);
+        self.get_unchecked(storage, pos)
+            .and_then(|v| v.ok_or_else(|| StdError::not_found(format!("deque position {}", pos))))
+            .map(Some)
+    }
+
+    /// Tries to get the value at the given position
+    /// Used internally
+    fn get_unchecked(&self, storage: &dyn Storage, pos: u32) -> StdResult<Option<T>> {
+        let prefixed_key = namespaces_with_key(&[self.namespace], &pos.to_be_bytes());
+        may_deserialize(&storage.get(&prefixed_key))
+    }
+
+    /// Removes the value at the given position
+    /// Used internally
+    fn remove_unchecked(&self, storage: &mut dyn Storage, pos: u32) {
+        let prefixed_key = namespaces_with_key(&[self.namespace], &pos.to_be_bytes());
+        storage.remove(&prefixed_key);
+    }
+
+    /// Tries to set the value at the given position
+    /// Used internally when pushing
+    fn set_unchecked(&self, storage: &mut dyn Storage, pos: u32, value: &T) -> StdResult<()> {
+        let prefixed_key = namespaces_with_key(&[self.namespace], &pos.to_be_bytes());
+        storage.set(&prefixed_key, &to_vec(value)?);
+
+        Ok(())
+    }
+}
+
+// used internally to avoid additional storage loads
+#[inline]
+fn calc_len(head: u32, tail: u32) -> u32 {
+    tail.wrapping_sub(head)
+}
+
+#[cfg(feature = "iterator")]
+impl<'a, T: Serialize + DeserializeOwned> VecDeque<'a, T> {
+    pub fn iter(&self, storage: &'a dyn Storage) -> StdResult<VecDequeIter<T>> {
+        Ok(VecDequeIter {
+            deque: self,
+            storage,
+            start: self.head(storage)?,
+            end: self.tail(storage)?,
+        })
+    }
+}
+
+#[cfg(feature = "iterator")]
+pub struct VecDequeIter<'a, T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    deque: &'a VecDeque<'a, T>,
+    storage: &'a dyn Storage,
+    start: u32,
+    end: u32,
+}
+
+#[cfg(feature = "iterator")]
+impl<'a, T> Iterator for VecDequeIter<'a, T>
```

Comment:
The range/iterator implementation is genius.

Do I understand correctly that this even works without the native iterator implementation? I think you can remove `#[cfg(feature = "iterator")]` since you don't rely on storage ranges.

Reply:
Yes, it only needs `storage.get`. Actually, I was wondering about that iterator feature. Are there CosmWasm chains that don't support `storage.range`?

Reply:
IIRC yes, there are chains that don't support it due to how their storage works. I'm sure Simon or Ethan could elaborate.

Reply:
There are underlying storage engines that don't have iterators. The default storage has it but e.g. Secret Network has an encrypted KV storage which you cannot iterate. Also Patricia Merkle Tries (Ethereum) do not support iterating.